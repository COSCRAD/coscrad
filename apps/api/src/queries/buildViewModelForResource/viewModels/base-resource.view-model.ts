import { ContributorWithId } from '@coscrad/api-interfaces';
import { NonEmptyString } from '@coscrad/data-types';
import { isNullOrUndefined } from '@coscrad/validation-constraints';
import idEquals from '../../../domain/models/shared/functional/idEquals';
import { CoscradContributor } from '../../../domain/models/user-management/contributor';
import { BaseViewModel, Nameable } from './base.view-model';
import { HasViewModelId } from './types/ViewModelId';

interface Accreditable {
    getContributions(): { contributorId: string; eventType: string; date: number }[];
}

export class BaseResourceViewModel extends BaseViewModel {
    @NonEmptyString({
        label: `contributions`,
        description: `list of knowledge keepers who contributed this song`,
        isArray: true,
    })
    readonly contributions: ContributorWithId[];

    constructor(
        domainModel: HasViewModelId & Nameable & Accreditable,
        contributors: CoscradContributor[]
    ) {
        super(domainModel);

        const contributionsWithDuplicates = domainModel
            .getContributions()
            .map(({ contributorId }) => contributors.find(idEquals(contributorId)))
            .filter((contributor) => !isNullOrUndefined(contributor))
            .map(({ id, fullName }) => ({ id: id, fullName: fullName.toString() }));

        this.contributions = [...new Set(contributionsWithDuplicates)];
    }
}
