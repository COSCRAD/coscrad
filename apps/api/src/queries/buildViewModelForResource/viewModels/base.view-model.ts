import { IBaseViewModel, IMultilingualText } from '@coscrad/api-interfaces';
import { FromDomainModel, NestedDataType } from '@coscrad/data-types';
import { ApiProperty } from '@nestjs/swagger';
import { MultilingualText } from '../../../domain/common/entities/multilingual-text';
import { Aggregate } from '../../../domain/models/aggregate.entity';
import { Ctor } from '../../../lib/types/Ctor';
import { HasViewModelId, ViewModelId } from './types/ViewModelId';

interface Nameable {
    getName(): MultilingualText;
}

export class BaseViewModel implements IBaseViewModel {
    readonly type: string;

    @ApiProperty({
        example: '12',
        description: 'uniquely identifies an entity from other entities of the same type',
    })
    @FromDomainModel(Aggregate as Ctor<unknown>)
    readonly id: ViewModelId;

    @NestedDataType(MultilingualText, {
        description: `multilingual text name of the entity`,
        label: `name`,
    })
    readonly name: IMultilingualText;

    constructor(domainModel: HasViewModelId & Nameable) {
        this.id = domainModel.id;

        this.name = domainModel.getName();
    }
}
