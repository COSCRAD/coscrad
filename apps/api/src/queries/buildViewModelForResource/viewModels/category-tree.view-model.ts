import { ICategoryTreeViewModel, ICompositeIdentifier } from '@coscrad/api-interfaces';
import { FromDomainModel, NestedDataType } from '@coscrad/data-types';
import { buildMultilingualTextWithSingleItem } from '../../../domain/common/build-multilingual-text-with-single-item';
import { Category } from '../../../domain/models/categories/entities/category.entity';
import { CategorizableType } from '../../../domain/types/CategorizableType';
import cloneToPlainObject from '../../../lib/utilities/cloneToPlainObject';
import { BaseViewModel } from './base.view-model';

import buildTreeFromNodes from './utilities/graph/buildTreeFromNodes';

const FromCategory = FromDomainModel(Category);

export class CategoryTreeViewModel
    extends BaseViewModel
    implements ICategoryTreeViewModel<CategorizableType>
{
    @FromCategory
    readonly label: string;

    @FromCategory
    readonly members: ICompositeIdentifier<CategorizableType>[];

    /**
     * We may need to rename our `FromDomainModel` decorator to
     * `FromRelatedClass` or `FromClass` or `From`.
     */
    /**
     * **WARNING** This property must be defind last!
     *
     * This works because of the order in which decorators are applied. Provided
     * that the `children` property is the last one defined, `NestedDataType`
     * will look up the current schema for `CategoryTreeViewModel` using
     * reflection metadata and find the complete definition.
     */
    @NestedDataType(CategoryTreeViewModel, {
        label: 'children',
        description:
            'self similar reference to all categories that are direct-descendants of this category in the tree',
    })
    // Note the use of the concrete type here
    readonly children: CategoryTreeViewModel[];

    constructor(categoryTree: Category[]) {
        const tree = buildTreeFromNodes(categoryTree);

        // TODO remove hack
        super({ id: tree.id, getName: () => buildMultilingualTextWithSingleItem(tree.label) });

        const { label, members, children } = tree;

        this.label = label;

        this.members = cloneToPlainObject(members);

        this.children = cloneToPlainObject(children);
    }
}
