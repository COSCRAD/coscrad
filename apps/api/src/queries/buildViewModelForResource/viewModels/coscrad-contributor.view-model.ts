import { ICoscradContributorViewModel } from '@coscrad/api-interfaces';
import { FromDomainModel, NonEmptyString } from '@coscrad/data-types';
import { isNullOrUndefined } from '@coscrad/validation-constraints';
import { CoscradContributor } from '../../../domain/models/user-management/contributor/entities/coscrad-contributor.entity';
import { BaseViewModel } from './base.view-model';

const FromContributor = FromDomainModel(CoscradContributor);

// TODO Annotate this as a COSCRAD canonical view
export class CoscradContributorViewModel
    extends BaseViewModel
    implements ICoscradContributorViewModel
{
    @NonEmptyString({
        label: 'full name',
        description: 'full name of the contributor',
    })
    fullName: string;

    @FromContributor
    shortBio?: string;

    constructor(contributor: CoscradContributor) {
        super(contributor);

        const { shortBio } = contributor;

        if (!isNullOrUndefined(shortBio)) {
            this.shortBio = shortBio;
        }

        this.fullName = contributor.getName().getOriginalTextItem().text;
    }
}
