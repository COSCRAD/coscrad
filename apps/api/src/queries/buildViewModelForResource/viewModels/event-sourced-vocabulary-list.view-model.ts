import {
    ContributorWithId,
    FormFieldType,
    ICommandFormAndLabels,
    IDetailQueryResult,
    IDynamicForm,
    IFormField,
    IMultilingualText,
    IVocabularyListEntry,
    IVocabularyListViewModel,
} from '@coscrad/api-interfaces';
import { ICoscradEvent } from '../../../domain/common';
import { buildMultilingualTextWithSingleItem } from '../../../domain/common/build-multilingual-text-with-single-item';
import { AccessControlList } from '../../../domain/models/shared/access-control/access-control-list.entity';
import {
    FilterPropertyType,
    VocabularyListCreated,
    VocabularyListFilterPropertyRegistered,
} from '../../../domain/models/vocabulary-list/commands';

export class EventSourcedVocabularyListViewModel
    implements IDetailQueryResult<IVocabularyListViewModel>
{
    entries: IVocabularyListEntry<string | boolean>[];
    form: IDynamicForm;
    contributions: ContributorWithId[];
    name: IMultilingualText;
    id: string;
    actions: ICommandFormAndLabels[];
    isPublished: boolean;
    accessControlList: { allowedUserIds: string[]; allowedGroupIds: string[] };

    static fromVocabularyListCreated({
        payload: {
            name: textForName,
            languageCodeForName,
            aggregateCompositeIdentifier: { id: vocabularyListId },
        },
    }: VocabularyListCreated): EventSourcedVocabularyListViewModel {
        const view = new EventSourcedVocabularyListViewModel();

        view.id = vocabularyListId;

        view.isPublished = false;

        view.entries = [];

        view.form = {
            fields: [],
        };

        view.contributions = [];

        // TODO we should serialize when returning from the query service automatically so we can use instances here if we'd like
        view.name = buildMultilingualTextWithSingleItem(textForName, languageCodeForName).toDTO();

        view.actions = [];

        view.accessControlList = new AccessControlList();

        return view;
    }

    /**
     * TODO If we really want to do this, we need to find a way to share the
     * logic with the event consumers. Unfortunately, the latter simply emit
     * deltas to the database with no up-front reads. It seems that we might
     * want a pattern where we calculate the delta to the view model for a given
     * event, and apply simply becomes pattern match + left fold over these deltas.
     *
     * For now, since the only use case is using event-sourced instead of state-based test setup,
     * we have decided to simpy change our approach to state based.
     */
    apply(event: ICoscradEvent) {
        if (event.isOfType('VOCABULARY_LIST_PROPERTY_FILTER_REGISTERED')) {
            const {
                payload: { name, type: filterType, allowedValuesAndLabels },
            } = event as VocabularyListFilterPropertyRegistered;

            const newFormField: IFormField<string | boolean> = {
                name,
                description: `filter the vocabulary list based on the property: ${name}`,
                // TODO ensure this comes through when writing to the database
                type:
                    filterType === FilterPropertyType.checkbox
                        ? FormFieldType.switch
                        : FormFieldType.staticSelect,

                label: name,
                // TODO should we add these?
                constraints: [],
                options: allowedValuesAndLabels.map(({ label, value }) => ({
                    value,
                    display: label,
                })),
            };

            this.form.fields.push(newFormField);

            return this;
        }

        // event not handled
        return this;
    }
}
