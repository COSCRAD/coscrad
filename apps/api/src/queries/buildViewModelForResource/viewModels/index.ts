export * from './playlist.view-model';
export * from './song.view-model';
export * from './tag.view-model';
export * from './term.view-model';
export * from './types/ViewModelId';
export * from './vocabulary-list.view-model';
