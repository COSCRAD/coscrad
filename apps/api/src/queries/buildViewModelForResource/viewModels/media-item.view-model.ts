import { IMediaItemViewModel } from '@coscrad/api-interfaces';
import { FromDomainModel, MIMEType } from '@coscrad/data-types';
import { MediaItem } from '../../../domain/models/media-item/entities/media-item.entity';
import { BaseViewModel } from './base.view-model';

const FromMediaItem = FromDomainModel(MediaItem);

export class MediaItemViewModel extends BaseViewModel implements IMediaItemViewModel {
    @FromMediaItem
    readonly url: string;

    @FromMediaItem
    readonly mimeType: MIMEType;

    @FromMediaItem
    readonly lengthMilliseconds: number;

    readonly filepath: string;

    constructor(mediaItem: MediaItem) {
        const { url, mimeType, lengthMilliseconds } = mediaItem;

        super(mediaItem);

        this.url = url;

        this.mimeType = mimeType;

        this.lengthMilliseconds = lengthMilliseconds;

        this.filepath = mediaItem.getFilePath();
    }
}
