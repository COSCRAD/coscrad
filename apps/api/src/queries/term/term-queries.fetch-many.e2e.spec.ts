import {
    AggregateType,
    CoscradUserRole,
    HttpStatusCode,
    LanguageCode,
    ResourceType,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes from '../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../app/controllers/__tests__/setUpIntegrationTest';
import getValidAggregateInstanceForTest from '../../domain/__tests__/utilities/getValidAggregateInstanceForTest';
import { ICoscradEvent, ICoscradEventHandler } from '../../domain/common';
import buildDummyUuid from '../../domain/models/__tests__/utilities/buildDummyUuid';
import { ResourceReadAccessGrantedToUser } from '../../domain/models/shared/common-commands';
import { ResourceReadAccessGrantedToUserEventHandler } from '../../domain/models/shared/common-commands/grant-resource-read-access-to-user/resource-read-access-granted-to-user.event-handler';
import { ResourcePublished } from '../../domain/models/shared/common-commands/publish-resource/resource-published.event';
import { ResourcePublishedEventHandler } from '../../domain/models/shared/common-commands/publish-resource/resource-published.event-handler';
import { TermCreated, TermTranslated } from '../../domain/models/term/commands';
import { AudioAddedForTermEventHandler } from '../../domain/models/term/commands/add-audio-for-term/audio-added-for-term.event-handler';
import { TermCreatedEventHandler } from '../../domain/models/term/commands/create-term/term-created.event-handler';
import { TermTranslatedEventHandler } from '../../domain/models/term/commands/translate-term/term-translated.event-handler';
import {
    ITermQueryRepository,
    TERM_QUERY_REPOSITORY_TOKEN,
} from '../../domain/models/term/queries';
import { CoscradUserGroup } from '../../domain/models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUserWithGroups } from '../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../domain/models/user-management/user/entities/user/coscrad-user.entity';
import { FullName } from '../../domain/models/user-management/user/entities/user/full-name.entity';
import { InternalError } from '../../lib/errors/InternalError';
import { ArangoDatabaseProvider } from '../../persistence/database/database.provider';
import TestRepositoryProvider from '../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import buildTestDataInFlatFormat from '../../test-data/buildTestDataInFlatFormat';
import { TestEventStream } from '../../test-data/events';
import { TermViewModel } from '../buildViewModelForResource/viewModels';

const indexEndpoint = `/resources/terms`;

// We require an existing user for ACL tests
const dummyQueryUserId = buildDummyUuid(4);

const { user: users, userGroup: userGroups } = buildTestDataInFlatFormat();

const dummyUser = (users as CoscradUser[])[0].clone({
    authProviderUserId: `auth0|${dummyQueryUserId}`,
    id: dummyQueryUserId,
    roles: [CoscradUserRole.viewer],
});

const dummyGroup = (userGroups as CoscradUserGroup[])[0].clone({ userIds: [dummyUser.id] });

const dummyUserWithGroups = new CoscradUserWithGroups(dummyUser, [dummyGroup]);

// Set up test data (use event sourcing to set up state)
const termText = `Term (in the language)`;

const originalLanguage = LanguageCode.Haida;

const termTranslation = `Term (in English)`;

const translationLanguage = LanguageCode.English;

const termId = buildDummyUuid(101);

const termIdUnpublishedNoUserAccessId = buildDummyUuid(102);

const termIdUnpublishedWithUserAccessId = buildDummyUuid(103);

const dummyContributorFirstName = 'Dumb';

const dummyContributorLastName = 'McContributor';

const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor).clone({
    fullName: new FullName({
        firstName: dummyContributorFirstName,
        lastName: dummyContributorLastName,
    }),
});

const termCreated = new TestEventStream().andThen<TermCreated>({
    type: 'TERM_CREATED',
    payload: {
        text: termText,
        languageCode: originalLanguage,
    },
    meta: {
        // we also want to check that the contributions come through
        contributorIds: [dummyContributor.id],
    },
});

const termTranslated = termCreated.andThen<TermTranslated>({
    type: 'TERM_TRANSLATED',
    payload: {
        translation: termTranslation,
        languageCode: translationLanguage,
    },
});

const termPrivateThatUserCanAccess = termTranslated.andThen<ResourceReadAccessGrantedToUser>({
    type: 'RESOURCE_READ_ACCESS_GRANTED_TO_USER',
    payload: {
        userId: dummyQueryUserId,
    },
});

const termPublished = termTranslated.andThen<ResourcePublished>({
    type: 'RESOURCE_PUBLISHED',
});

// const promptTermId = buildDummyUuid(2)

/**
 * TODO We need to find a more maintainable way of
 * seeding the required initial state.
 */
const buildEventHandlers = (termQueryRepository: ITermQueryRepository) => [
    new TermCreatedEventHandler(termQueryRepository),
    new TermTranslatedEventHandler(termQueryRepository),
    new AudioAddedForTermEventHandler(termQueryRepository),
    // TODO update this to take in a generic query repository provider
    new ResourceReadAccessGrantedToUserEventHandler(termQueryRepository),
    new ResourcePublishedEventHandler({
        // TODO break this out into an ArangoQueryRepositoryProvider
        forResource(resourceType: ResourceType) {
            if (resourceType === ResourceType.term) {
                return termQueryRepository;
            }

            throw new InternalError(
                `Resource Type: ${resourceType} is not supported by the query repository provider`
            );
        },
    }),
];

describe(`when querying for a term: fetch many`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let termQueryRepository: ITermQueryRepository;

    let databaseProvider: ArangoDatabaseProvider;

    let handlers: ICoscradEventHandler[];

    let seedTerms: (eventHistory: ICoscradEvent[]) => Promise<void>;

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();

        await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    const eventHistoryForManyUnpublished = [
        ...termTranslated.as({ id: termIdUnpublishedNoUserAccessId, type: AggregateType.term }),
        ...termPrivateThatUserCanAccess.as({
            id: termIdUnpublishedWithUserAccessId,
            type: AggregateType.term,
        }),
    ];

    const eventHistoryForManyWithPublishedTerm = [
        ...eventHistoryForManyUnpublished,
        ...termPublished.as({ id: termId, type: AggregateType.term }),
    ];

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            await app.init();

            termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

            handlers = buildEventHandlers(termQueryRepository);

            seedTerms = async (events: ICoscradEvent[]) => {
                for (const e of events) {
                    for (const h of handlers) {
                        await h.handle(e);
                    }
                }
            };
        });

        describe(`when there is a published term in the index view`, () => {
            it(`should only return published terms`, async () => {
                // TODO -> beforeEach
                await seedTerms(eventHistoryForManyWithPublishedTerm);

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);

                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                const {
                    body: { entities },
                } = res;

                // TODO check that the actions are correct !!!!!
                expect(entities).toHaveLength(1);

                // Only one published result should come through from eventHistoryForMany
                const result = entities[0] as TermViewModel;

                expect(result.id).toBe(termId);

                expect(
                    result.contributions.some(({ fullName, id: foundContributorId }) => {
                        return (
                            // TODO we will want a test helper for this comparison in case we decide to change format
                            fullName ===
                                `${dummyContributorFirstName} ${dummyContributorLastName}` &&
                            foundContributorId === dummyContributor.id
                        );
                    })
                ).toBe(true);
            });
        });

        describe(`when there are no published terms`, () => {
            it(`should return no terms`, async () => {
                // -> beforeEach
                await seedTerms(eventHistoryForManyUnpublished);

                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                expect(res.body.entities).toHaveLength(0);
            });
        });
    });

    describe(`when the user is authenticated as a non-admin user`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                },
                {
                    testUserWithGroups: dummyUserWithGroups,
                }
            ));

            termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

            handlers = buildEventHandlers(termQueryRepository);

            seedTerms = async (events: ICoscradEvent[]) => {
                for (const e of events) {
                    for (const h of handlers) {
                        await h.handle(e);
                    }
                }
            };
        });

        describe(`when there is a term that is unpublished`, () => {
            describe(`when the user does not have read access`, () => {
                it(`should not return the unpublished term`, async () => {
                    // -> beforeEach
                    await seedTerms(eventHistoryForManyWithPublishedTerm);

                    const res = await request(app.getHttpServer()).get(indexEndpoint);

                    expect(res.status).toBe(httpStatusCodes.ok);

                    /**
                     * + private, but user in ACL
                     * - private, user not in ACL
                     * + published
                     */
                    expect(res.body.entities).toHaveLength(2);
                });
            });

            describe(`when the user does have read access`, () => {
                it(`should return the unpublished term`, async () => {
                    await seedTerms(eventHistoryForManyWithPublishedTerm);

                    const res = await request(app.getHttpServer()).get(indexEndpoint);

                    expect(res.status).toBe(httpStatusCodes.ok);

                    /**
                     * + published
                     * + private, but user in ACL
                     * - private, user not in ACL
                     */

                    const {
                        body: { entities },
                    } = res;

                    const result = entities[0] as TermViewModel;

                    expect(result.id).toBe(termIdUnpublishedWithUserAccessId);
                });
            });
        });
    });

    (
        [
            [CoscradUserRole.superAdmin, 'when the user is a COSCRAD admin'],
            [CoscradUserRole.projectAdmin, 'when the user is a project admin'],
        ] as const
    ).forEach(([userRole, description]) => {
        describe(description, () => {
            beforeAll(async () => {
                ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                    {
                        ARANGO_DB_NAME: testDatabaseName,
                    },
                    {
                        testUserWithGroups: new CoscradUserWithGroups(
                            dummyUser.clone({
                                roles: [userRole],
                            }),
                            []
                        ),
                    }
                ));

                termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

                handlers = buildEventHandlers(termQueryRepository);

                seedTerms = async (events: ICoscradEvent[]) => {
                    for (const e of events) {
                        for (const h of handlers) {
                            await h.handle(e);
                        }
                    }
                };
            });

            it(`should allow the user to access private resources`, async () => {
                // -> beforeEach
                await seedTerms(eventHistoryForManyUnpublished);

                const res = await request(app.getHttpServer()).get(indexEndpoint);

                const numberOfPrivateTerms = 2;

                expect(res.status).toBe(HttpStatusCode.ok);

                expect(res.body.entities).toHaveLength(numberOfPrivateTerms);

                /**
                 * This if statement is effectively a filter for our test
                 * case builder pattern. We only want to snapshot the response
                 * once for one specific class of user. We want this to be
                 * an admin user, because admin users have access to all
                 * data, including command info (available actions).
                 * Finally, we do this for the project admin, as queries
                 * for this user are by far more common than for coscrad admin
                 * in practice.
                 */
                if (userRole === CoscradUserRole.projectAdmin) {
                    /**
                     * This is to catch breaking changes in the API contract with
                     * the client. See the above comment for the corresponding detail
                     * endpoint test case.
                     */
                    expect(res.body).toMatchSnapshot();
                }
            });
        });
    });
});
