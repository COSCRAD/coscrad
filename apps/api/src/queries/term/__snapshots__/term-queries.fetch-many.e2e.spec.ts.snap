// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a term: fetch many when the user is a project admin should allow the user to access private resources 1`] = `
{
  "entities": [
    {
      "actions": [
        {
          "description": "Allow a user to view (but not edit) a given resource",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a UUID",
                    "name": "UUID",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "the ID of the user who will be given permission to view this resource",
                "label": "userId",
                "name": "userId",
                "options": {
                  "aggregateType": "user",
                },
                "type": "DYNAMIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "term",
              },
            },
          },
          "label": "Grant Read Access to User",
          "type": "GRANT_RESOURCE_READ_ACCESS_TO_USER",
        },
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "term",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "Translate an existing term (typically from the language to English)",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "translation for the given term",
                "label": "translation",
                "name": "translation",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "language in which you are translating the term (typically English)",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "term",
              },
            },
          },
          "label": "Translate Term",
          "type": "TRANSLATE_TERM",
        },
      ],
      "contributions": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Term (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Term (in English)",
          },
        ],
      },
      "tags": [],
    },
    {
      "actions": [
        {
          "description": "Allow a user to view (but not edit) a given resource",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a UUID",
                    "name": "UUID",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "the ID of the user who will be given permission to view this resource",
                "label": "userId",
                "name": "userId",
                "options": {
                  "aggregateType": "user",
                },
                "type": "DYNAMIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "term",
              },
            },
          },
          "label": "Grant Read Access to User",
          "type": "GRANT_RESOURCE_READ_ACCESS_TO_USER",
        },
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "term",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "Translate an existing term (typically from the language to English)",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "translation for the given term",
                "label": "translation",
                "name": "translation",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "language in which you are translating the term (typically English)",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "term",
              },
            },
          },
          "label": "Translate Term",
          "type": "TRANSLATE_TERM",
        },
      ],
      "contributions": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Term (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Term (in English)",
          },
        ],
      },
      "tags": [],
    },
  ],
  "indexScopedActions": [
    {
      "description": "Create a new term in the language",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the term (in the language)",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language of this term",
            "label": "language code",
            "name": "languageCode",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Term",
      "type": "CREATE_TERM",
    },
    {
      "description": "Create a new Prompt Term in the language",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the term (in English)",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Prompt Term",
      "type": "CREATE_PROMPT_TERM",
    },
  ],
}
`;
